sdk/lib/rtl/heap.c

```diff
@@ -426,6 +434,10 @@ RtlpRemoveFreeBlock(PHEAP Heap,
     SIZE_T Result, RealSize;
     ULONG HintIndex;
 
+    /* This was a problem before we handled segments > MAXUSHORT.
```

Comment:
What do you mean here? We are able to handle segment sizes larger than MAXUSHORT now?

However, how is this possible? The [`HEAP_FREE_ENTRY::Size`](https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/lib/rtl/heap.h;hb=edf6b805a7fee8ca2f1dec9d97acae510f4f6082#l129) member comes from the [`struct _HEAP_COMMON_ENTRY`](https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/lib/rtl/heap.h;hb=edf6b805a7fee8ca2f1dec9d97acae510f4f6082#l77) structure, which is typed as `USHORT` , so any value stored inside it cannot go over `MAXUSHORT` .

Another observation is that in e.g. the `RtlpInsertFreeBlock()` function there are checks for heap blocks whose chunk size shouldn't go over `HEAP_MAX_BLOCK_SIZE`. You can verify by calculating the value of the [`HEAP_MAX_BLOCK_SIZE` define](https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/lib/rtl/heap.h;hb=edf6b805a7fee8ca2f1dec9d97acae510f4f6082#l16)  that this value is < `MAXUSHORT`.

Reply:
> What do you mean here? We are able to handle segment sizes larger than MAXUSHORT now?

Yes, ReactOS should now be able to handle this now by using multiple smaller segments that are "chained". That is the reason that ```USHORT``` and other smaller sized variables are OK here. Thanks.